PriorityElectionTimerTest#shouldHighPriorityNodeStartElectionFirst is flaky

**Summary**

> Try to answer the following as best as possible

- How often does the test fail? at least 3 times in the last week
- Does it block your work? no
- Do we suspect that it is a real failure? no, I think it's a test issue, but not familiar enough at a glance

**Failures**

<details><summary>Example assertion failure</summary>
<pre>
[2021-12-12T00:12:31.927Z] [ERROR] io.atomix.raft.impl.PriorityElectionTimerTest.shouldHighPriorityNodeStartElectionFirst  Time elapsed: 3.475 s  <<< FAILURE!
[2021-12-12T00:12:31.927Z] org.mockito.exceptions.verification.VerificationInOrderFailure: 
[2021-12-12T00:12:31.927Z] 
[2021-12-12T00:12:31.927Z] Verification in order failure:
[2021-12-12T00:12:31.927Z] atomicBoolean.set(true);
[2021-12-12T00:12:31.927Z] Wanted 1 time:
[2021-12-12T00:12:31.927Z] -> at java.base/java.util.concurrent.atomic.AtomicBoolean.set(AtomicBoolean.java:153)
[2021-12-12T00:12:31.927Z] But was 3 times:
[2021-12-12T00:12:31.927Z] -> at io.atomix.raft.impl.PriorityElectionTimerTest.lambda$shouldHighPriorityNodeStartElectionFirst$1(PriorityElectionTimerTest.java:68)
[2021-12-12T00:12:31.927Z] -> at io.atomix.raft.impl.PriorityElectionTimerTest.lambda$shouldHighPriorityNodeStartElectionFirst$1(PriorityElectionTimerTest.java:68)
[2021-12-12T00:12:31.927Z] -> at io.atomix.raft.impl.PriorityElectionTimerTest.lambda$shouldHighPriorityNodeStartElectionFirst$1(PriorityElectionTimerTest.java:68)
[2021-12-12T00:12:31.927Z] 
</pre>
</details>

**Hypotheses**

I couldn't reproduce it locally after running it thousands of time, but I suspect it could be some race condition where one timer ends up being called more often than the other due to thread scheduling issues.

--- a/atomix/cluster/src/test/java/io/atomix/raft/impl/PriorityElectionTimerTest.java
+++ b/atomix/cluster/src/test/java/io/atomix/raft/impl/PriorityElectionTimerTest.java
@@ -23,23 +23,23 @@ import java.util.List;
 import java.util.concurrent.CopyOnWriteArrayList;
 import java.util.concurrent.atomic.AtomicInteger;
 import org.awaitility.Awaitility;
-import org.junit.After;
-import org.junit.Test;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.Test;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-public class PriorityElectionTimerTest {
+final class PriorityElectionTimerTest {
 
   private final Logger log = LoggerFactory.getLogger(PriorityElectionTimerTest.class);
   private final SingleThreadContext threadContext = new SingleThreadContext("priorityElectionTest");
 
-  @After
-  public void after() {
+  @AfterEach
+  void afterEach() {
     threadContext.close();
   }
 
   @Test
-  public void shouldLowerPriorityNodeEventuallyStartsAnElection() {
+  void shouldLowerPriorityNodeEventuallyStartsAnElection() {
     // given
     final AtomicInteger triggerCount = new AtomicInteger();
 
@@ -55,7 +55,7 @@ public class PriorityElectionTimerTest {
   }
 
   @Test
-  public void shouldHighPriorityNodeStartElectionFirst() {
+  void shouldHighPriorityNodeStartElectionFirst() {
     // given
     final String highPrioId = "highPrioTimer";
     final String lowPrioId = "lowPrioTimer";

--- a/atomix/cluster/src/test/java/io/atomix/raft/impl/PriorityElectionTimerTest.java
+++ b/atomix/cluster/src/test/java/io/atomix/raft/impl/PriorityElectionTimerTest.java
@@ -15,16 +15,16 @@
  */
 package io.atomix.raft.impl;
 
-import static org.mockito.Mockito.spy;
+import static org.assertj.core.api.Assertions.assertThat;
 
 import io.atomix.utils.concurrent.SingleThreadContext;
 import java.time.Duration;
-import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.List;
+import java.util.concurrent.CopyOnWriteArrayList;
 import java.util.concurrent.atomic.AtomicInteger;
 import org.awaitility.Awaitility;
 import org.junit.After;
 import org.junit.Test;
-import org.mockito.Mockito;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -57,15 +57,16 @@ public class PriorityElectionTimerTest {
   @Test
   public void shouldHighPriorityNodeStartElectionFirst() {
     // given
-    final AtomicBoolean highPrioElectionTriggered = spy(new AtomicBoolean());
-    final AtomicBoolean lowPrioElectionTriggered = spy(new AtomicBoolean());
+    final String highPrioId = "highPrioTimer";
+    final String lowPrioId = "lowPrioTimer";
+    final List<String> electionOrder = new CopyOnWriteArrayList<>();
 
     final int targetPriority = 4;
     final PriorityElectionTimer timerHighPrio =
         new PriorityElectionTimer(
             Duration.ofMillis(100),
             threadContext,
-            () -> highPrioElectionTriggered.set(true),
+            () -> electionOrder.add(highPrioId),
             log,
             targetPriority,
             targetPriority);
@@ -74,7 +75,7 @@ public class PriorityElectionTimerTest {
         new PriorityElectionTimer(
             Duration.ofMillis(100),
             threadContext,
-            () -> lowPrioElectionTriggered.set(true),
+            () -> electionOrder.add(lowPrioId),
             log,
             targetPriority,
             1);
@@ -82,12 +83,16 @@ public class PriorityElectionTimerTest {
     // when
     timerLowPrio.reset();
     timerHighPrio.reset();
-    Awaitility.await().until(highPrioElectionTriggered::get);
-    Awaitility.await().until(lowPrioElectionTriggered::get);
 
     // then
-    final var inorder = Mockito.inOrder(highPrioElectionTriggered, lowPrioElectionTriggered);
-    inorder.verify(highPrioElectionTriggered).set(true);
-    inorder.verify(lowPrioElectionTriggered).set(true);
+    Awaitility.await()
+        .untilAsserted(
+            () ->
+                assertThat(electionOrder)
+                    .as("both elections should have been triggered eventually")
+                    .contains(highPrioId, lowPrioId));
+    assertThat(electionOrder.get(0))
+        .as("the first election triggered should have been the high priority election")
+        .isEqualTo(highPrioId);
   }
 }

